### ISSUE TITLE
Change Pedalboard via MIDI

### ISSUE THREAD

**Issue by gallochri** [TreeFallSound/pi-stomp#9]
Is it possible to change pedalboard via midi?

**Comment by rreichenbach** [comment]
Yes.  Program Change messages coming into the MIDI port should change the pedalboard active in MOD.  The bad news is that the pi-Stomp software (separate from MOD) doesn't currently listen to incoming MIDI messages, so the LCD will not reflect the pedalboard change.  It certainly could.  Feel free to submit an issue to request adding that functionality.

**Comment by gallochri** [comment]
I can't find which channel to send "Program Change" messages to change pedalboards.
Is there a log of incoming midi messages?

**Comment by rreichenbach** [comment]
pi-stomp specific software doesn't log messages coming in, only the messages it generates.  You could likely use a midi utility like amidi with the dump option, but that doesn't seem to work for me.  Also, I'm not seeing Mod respond to PC messages.  It did at one time, but it seemed they might have changed that or there's some setting required to enable it.

**Comment by ozbenh** [comment]
Talking of which, when the board is changed via the web UI, this is also not reflected in pi-stomp... should we also expect a MIDI message ? Otherwise is there a way to poll MOD for the current board and adapt ?

**Comment by rreichenbach** [comment]
There is no MIDI message sent when the pedalboard is changed via MOD UI.   And the communication between pi-stomp software and mod-ui/mod-host has been unidirectional via the rest API defined by mod-ui/webserver.py (woefully incomplete BTW).  For their own hardware, MOD does register callbacks for keeping the hardware in sync with the MOD UI.  I've had plans to do something similar, but haven't invested the time.  But yes, it's not very nice that at least pedalboard and preset changes from MOD UI are not synced on the pi-stomp.   A few months ago, I did create a polling based solution for pedalboard changes.  I've not discovered a similar hook for preset/snapshot changes, but I suppose pedalboard changes would be better than nothing.  I'll find and post that change once I find the SD card that has it (never pushed it).

**Comment by MacFizz** [comment]
In case anyone still interested in this, I've forked the project and worked on a plugin to change snapshots with MIDI PC and CC.
You can then configure whatever midi channel you're listening to with PC or CC to load the previous snapshot, the next or direcly a snapshot index.
Anyone care to test it ?

**Comment by rreichenbach** [comment]
Very cool!  Yes, would love to try it out (and merge it in eventually).  Please let us know how to use it.  Thanks.

**Comment by MacFizz** [comment]
Forked here: https://github.com/MacFizz/pi-stomp
It's creating one thread per midi device and monitor it until it finds corresponding pattern in the config file.
midi_snapshot_plugin.yml should be at the same location than your default_config.yml
Example file provided has wrong comments but is sufficiently simple and clear to be understood.
Please note, python is not my specialty. I've done my best not to break anything.
I also tested it with several USB controller (such as the M-Vave Chocolate) and works when connect in USB (no luck making it work on bluetooth for now)

**Comment by rreichenbach** [comment]
@MacFizz This is clever.  Thanks for sharing.  It looks like your plugin uses the old Mod handler mod.py not the new modhandler.py.  Things have changed significantly since then (see pistomp-v3 branch).  It shouldn't be too difficult to make it work with modhandler (both Mod and Modhandler inherit from the same Handler class), but I see a problem with threading.  Your plugin calls modapi.load_snapshot(index) and that will cause LCD changes.  LCD changes are managed in a different event loop, so asynchronous updates can cause simultaneous conflicts which will cause the LCD to blank.  A safer approach would probably be to queue the snapshot/pedalboard change, then let the synchronous LCD handling update on it's next cycle.  I might try prototyping something like that soon.
